common/s3util.h

```diff
@@ -187,7 +189,10 @@ class S3Util {
   }
 
   ~S3Util() {
-    Aws::ShutdownAPI(options_);
+    --counter_;
+    if (counter_.load() == 0) {
```

Comment:
This is a known issue with the current interface design. Adding an atomic reference count couldn't solve the problem. Imagine thread A finds load() returns 0, then thread B creates a new S3Util object, and then thread A calls ShutdownAPI().

A quick and dirty fix it is to make S3Util::S3Util() private, and merge the code https://github.com/pinterest/rocksplicator/blob/master/rocksdb_admin/admin_handler.cpp#L679-L712 into the static function BuildS3Util().

Reply:
You are right, the atomic counter does not solve the problem completely. The code you linked does not solve our problem either. Let's sync on Monday for a better plan. Thanks for the review.

Reply:
That's why "Can you describe your use case and the exact problem you want to fix?". As S3util can be used in various ways to trigger the bug, the code linked is how we managed to get around the problem.